### PR TITLE
Feature/pytorch refactoring

### PR DESCRIPTION
**Pull request by Cypher42**
Removed unused features
CheckLayerSat now searches recursivly through modules in order to find Linear and Conv2D layers
Made tensorboard optional. The user may now choose between print, csv and tensorboard output.
Increased performance bei only applying PCA when the metric is recorded.
The interface is now resembling MLFlow and Sacred more closely than before.

### CODE REVIEW

File: delve/torchcallback.py [hunk 1]
```diff
@@ -297,47 +300,62 @@ def record_layer_saturation(layer: torch.nn.Module, input, output):
...
 
         layer.register_forward_hook(record_layer_saturation)
+
+
+    def add_saturations(self):
+        """
+        Computes saturation and saves all stats
+        :return:
+        """
+        for key in self.logs.keys():
```
Comment (JohanByttner) [inline delve/torchcallback.py]: "for key in self.logs" would suffice

File: delve/torchcallback.py [hunk 2]
```diff
@@ -297,47 +300,62 @@ def record_layer_saturation(layer: torch.nn.Module, input, output):
...
+            train_sats = []
+            val_sats = []
+            if '-saturation' in key:
+                for layer_name in self.logs[key].keys():
+                    if not self.ignore_layer_names is None and self.ignore_layer_names == layer_name:
+                        continue
+                    sat = compute_saturation(self.logs[key][layer_name]._cov_mtx, thresh=.99)
```
Comment (JohanByttner) [inline delve/torchcallback.py]: the threshold should be straightforward to configure from outside the class

Comment (Cypher42) [reply delve/torchcallback.py]: done

File: delve/torchcallback.py (hunk 1 above)
Comment (Cypher42) [reply delve/torchcallback.py]: changed

File: delve/torchcallback.py [hunk 3]
```diff
@@ -297,47 +300,62 @@ def record_layer_saturation(layer: torch.nn.Module, input, output):
...
+        """
+        Computes saturation and saves all stats
+        :return:
+        """
+        for key in self.logs.keys():
+            train_sats = []
+            val_sats = []
+            if '-saturation' in key:
+                for layer_name in self.logs[key].keys():
+                    if not self.ignore_layer_names is None and self.ignore_layer_names == layer_name:
```
[the remark this answers is not in the document]
Comment (Cypher42) [reply delve/torchcallback.py]: switched to a empty list as a default, makes the first part obsolete
